### DISCUSSION TITLE
How do I proceed with inter-related dependency?

### DISCUSSION THREAD

**Discussion by Ashvith10** [Oldes/Rebol3#82] [General]
I'm trying to package Rebol3 on NixOS by creating an expression, but it looks like the build tool is not the already-available-and-supported `make` or `meson`. In order to build Rebol3, I need the Siskin-framework, but in order to use Siskin-framework, I also need Rebol3? I can't use binaries directly on NixOS, as you know, patchELF issues.

**Comment by Oldes** [comment]
Hi.. I've never heard about NixOS so I am not sure, what exactly do you need.. but if you want to compile Rebol (my fork) for a new system, you can use this [bootstrap](https://github.com/Siskin-framework/Rebol/tree/bootstrap) source.

Rebol sources are partially generated using Rebol itself.. the mentioned boostrap branch contains these generated files so one can just edit the makefile and make the initial version for the new OS.. and use this bootstrap version to make a real version.

**Comment by Oldes** [comment]
I just installed NixOS on my Mac and I can use Rebol there:
<img width="561" alt="Screenshot 2023-11-27 at 14 39 57" src="https://github.com/Oldes/Rebol3/assets/454805/b0abaa52-c90e-4ba1-9ed1-fa41af2b1733">

**Reply by Ashvith10** [reply]
@Oldes looks like there's some misunderstanding. I'll try to clear this to the best of my information:
- What you're using actually isn't NixOS, it is the Nix package manager. Think of it like a portable Homebrew (not exactly, but I'm generalizing it very much). As opposed to a system-level package manager, these are what we can call as a user-level package manager (it is also a system-level package manager, but that isn't relevant to platforms other that NixOS). NixOS is a reproducible, functional, transactional distro that uses the Nix package manager to build itself, using a Nix expression - you write a functional language called Nix-lang (or simply Nix) to build your Linux desktop, if that makes sense.
- Normal binaries do not work on NixOS, as opposed to MacOS or your regular Linux distro, as they're not stored in your typical root directory (`/usr`, `/share`, etc). They're instead, stored in `/nix/store` - leading to issues with dynamic loader.
- Also, because NixOS is functional, circular dependencies can be an issue.
- Now, this is not to say that you should have installed NixOS on a VM or a seperate machine, you can totally do almost all the stuff on NixOS over Nix, as long as the package specs allow it (`darwin` or `unix` packages will be available on Mac systems).
- What you've opened is a `nix-shell`, to be precise, an impure nix-shell. Now, what that means is that you're able to share both your local environment, as well as the `nix-shell` environment. You might want to try running it in pure mode - actually, you can't for now, as neither rebol, nor sisystem-framework exists in the package manager. However, there's a fork of Rebol called Red, which you can try out. But, hypothetically, if Rebol were a package in the stable version, the command would be something like `nix-shell -p rebol3 --pure`.

**Comment by acook** [comment]
NixOS is just a Linux distro with the Nix package manager as its only way to manage binaries and installs. 

The bootstrap branch should allow you to build a pipeline to generate the Rebol binary you need without a circular dependency.

**Comment by Oldes** [comment]
I should probably update the bootstrap branch as it is quite behind now (3.7.2).

Anyway... I am not familiar with all kinds of Linux packagers. I've tried to make a package for Alpine Linux some time ago, which was also the reason for the bootstrap branch, but I [give up that afford](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/44734#note_317218).

If I can help somehow with the Nix, just let me know how.

**Reply by Ashvith10** [reply]
@Oldes I'm using the bootstrap branch, just as you've mentioned, but I was wondering how to continue forward?

I've built the 64-bit version using `make -f rebol-linux-bootstrap-64bit.mk`, but it creates a folder `tmp` in the `make` directory, and not a binary?

```console
$ tree tmp
tmp
`-- gcc-64bit
    `-- rebol-linux-bootstrap-64bit
        `-- src
            |-- core
            |   |-- a-constants.o
            |   |-- a-globals.o
            |   |-- a-lib.o
            |   |-- b-boot.o
            |   |-- b-init.o
            |   |-- c-do.o
            |   |-- c-error.o
            |   |-- c-frame.o
            |   |-- c-function.o
            |   |-- c-handle.o
            |   |-- c-port.o
            |   |-- c-task.o
            |   |-- c-word.o
            |   |-- d-crash.o
            |   |-- d-dump.o
            |   |-- d-print.o
            |   |-- f-blocks.o
            |   |-- f-deci.o
            |   |-- f-dtoa.o
            |   |-- f-enbase.o
            |   |-- f-extension.o
            |   |-- f-int.o
            |   |-- f-math.o
            |   |-- f-modify.o
            |   |-- f-qsort.o
            |   |-- f-random.o
            |   |-- f-round.o
            |   |-- f-series.o
            |   |-- f-stubs.o
            |   |-- l-scan.o
            |   |-- l-types.o
            |   |-- m-gc.o
            |   |-- m-pools.o
            |   |-- m-series.o
            |   |-- mbedtls
            |   |   |-- md5.o
            |   |   |-- platform.o
            |   |   |-- platform_util.o
            |   |   |-- sha1.o
            |   |   |-- sha256.o
            |   |   `-- sha512.o
            |   |-- n-control.o
            |   |-- n-crypt.o
            |   |-- n-data.o
            |   |-- n-io.o
            |   |-- n-loop.o
            |   |-- n-math.o
            |   |-- n-sets.o
            |   |-- n-strings.o
            |   |-- n-system.o
            |   |-- p-checksum.o
            |   |-- p-console.o
            |   |-- p-dir.o
            |   |-- p-dns.o
            |   |-- p-event.o
            |   |-- p-file.o
            |   |-- p-net.o
            |   |-- s-cases.o
            |   |-- s-crc.o
            |   |-- s-file.o
            |   |-- s-find.o
            |   |-- s-make.o
            |   |-- s-mold.o
            |   |-- s-ops.o
            |   |-- s-trim.o
            |   |-- s-unicode.o
            |   |-- t-bitset.o
            |   |-- t-block.o
            |   |-- t-char.o
            |   |-- t-datatype.o
            |   |-- t-date.o
            |   |-- t-decimal.o
            |   |-- t-event.o
            |   |-- t-function.o
            |   |-- t-gob.o
            |   |-- t-handle.o
            |   |-- t-image.o
            |   |-- t-integer.o
            |   |-- t-logic.o
            |   |-- t-map.o
            |   |-- t-money.o
            |   |-- t-none.o
            |   |-- t-object.o
            |   |-- t-pair.o
            |   |-- t-port.o
            |   |-- t-string.o
            |   |-- t-struct.o
            |   |-- t-time.o
            |   |-- t-tuple.o
            |   |-- t-typeset.o
            |   |-- t-utype.o
            |   |-- t-vector.o
            |   |-- t-word.o
            |   |-- u-aes.o
            |   |-- u-bigint.o
            |   |-- u-bincode.o
            |   |-- u-chacha20.o
            |   |-- u-compress.o
            |   |-- u-dh.o
            |   |-- u-iconv.o
            |   |-- u-mbedtls.o
            |   |-- u-parse.o
            |   |-- u-poly1305.o
            |   |-- u-rc4.o
            |   |-- u-rsa.o
            |   |-- u-uECC.o
            |   `-- u-zlib.o
            `-- os
                |-- dev-dns.o
                |-- dev-net.o
                |-- host-args.o
                |-- host-device.o
                |-- host-main.o
                |-- host-stdio.o
                `-- posix
                    |-- dev-event.o
                    |-- dev-file.o
                    |-- dev-stdio.o
                    |-- host-lib.o
                    `-- host-readline.o
```

**Reply by Oldes** [reply]
The binary should be in the same folder like the make file (if the compilation is successful).

**Reply by Ashvith10** [reply]
@Oldes my bad, I should have checked it properly. I was able to build Rebol3 successfully.

![image](https://github.com/Oldes/Rebol3/assets/113123021/a855b028-3d13-44b6-b099-5f4e969707d9)

Just a suggestion - would you be okay with renaming the tag to `<version>-bootstrap` instead of `<version>-bootstrap-alpine`? Also, can you add tags for the new bootstrap version?

**Reply by Oldes** [reply]
Done: https://github.com/Siskin-framework/Rebol/releases/tag/3.15.0-bootstrap
Includes also the workflow for the bootstrap compilation:
https://github.com/Siskin-framework/Rebol/actions/runs/7035646575/job/19146496675

**Comment by Oldes** [comment]
The bootstrap branch was updated to version 3.15.0.

**Reply by Ashvith10** [reply]
@Oldes looks like this is going to be a lot of work than I had originally thought. Since Nix tries to create a reproducible build, stuff like git simply does not work inside the build. There is a sample example used by [red](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/interpreters/red/default.nix), but since it does not use seem to be using Siskin-framework, I'm gonna have to ask to a senior maintainer.

**Reply by Oldes** [reply]
You can get sources directly from any release without git... like: https://github.com/Oldes/Rebol3/archive/refs/tags/3.15.0.zip
Then you just need the build script, which is also provided with the Builder release, like this:
https://github.com/Siskin-framework/Builder/releases/download/0.13.0/siskin.r3.gz

(the Builder is just a customised Rebol which executes the script)

So you can just get the script and use it with the bootstrapped Rebol and the [rebol3.nest](https://github.com/Oldes/Rebol3/blob/master/make/rebol3.nest) file.

So it would be something like:
```
rebol-boostrap siskin.r3 Rebol3-3.15.0/make/rebol3.nest rebol3-bulk-linux-x64
```

**Reply by Oldes** [reply]
Of course.. it would be easier, when there would be already Rebol (or even Siskin) package as it is the only dependency.. except the compiler itself.

**Reply by Oldes** [reply]
To get the source and the script, it would be something like:
```
{ lib
, stdenv
, fetchurl
}:

stdenv.mkDerivation rec {
  pname = "rebol3";   # or just rebol?
  version = "3.15.0";

  src = fetchurl {
    url = "https://github.com/Oldes/Rebol3/archive/refs/tags/#{version}.tar.gz";
    sha256 = "sha256-bpnnBD69Gi8pThXOPJDIwfmaz429kZ5SVty9lskv6kI=";
  };

  siskin = fetchurl {
    url = "https://github.com/Siskin-framework/Builder/releases/download/0.13.0/siskin.r3.gz";
    sha256 = "sha256-TZEqOL1zJj7dT6jGoi5vsvHoaQpQUgXtzR4idU5S1GU=";
  };

  #...
}

```

**Reply by Ashvith10** [reply]
This is going to take a while. Since I'm a little busy now, I'll update you around next week.
